src/java/org/apache/cassandra/utils/bytecomparable/ByteComparable.md

```diff
@@ -322,104 +324,160 @@ To more efficiently encode byte-ordered blobs, however, we use a slightly tweake
 as an unsigned integer has the same effect as comparing the source signed.
 
 Examples:
 
-|Type and value|bytes|encodes as|
-|--------------|-----|----------|
-|int 1         |00 00 00 01|             80 00 00 01
-|short -1      |FF FF      |             7F FF
-|byte 0        |00         |             80
-|int MAX_VALUE |7F FF FF FF|             FF FF FF FF
-|long MIN_VALUE|80 00 00 00 00 00 00 00| 00 00 00 00 00 00 00 00
+
+| Type and value | bytes                   | encodes as              |
+| -------------- | ----------------------- | ----------------------- |
+| int 1          | 00 00 00 01             | 80 00 00 01             |
+| short -1       | FF FF                   | 7F FF                   |
+| byte 0         | 00                      | 80                      |
+| byte -2        | FE                      | 7E                      |
+| int MAX_VALUE  | 7F FF FF FF             | FF FF FF FF             |
+| long MIN_VALUE | 80 00 00 00 00 00 00 00 | 00 00 00 00 00 00 00 00 |
+
+## Variable-length encoding of integers (current bigint)
+
+Another way to encode integers that may save significant amounts of space when smaller numbers are often in use, but
+still permits large values to be efficiently encoded, is to use an encoding scheme similar to UTF-8.
+
+For unsigned numbers this can be done by starting the number with as many 1s in most significant bits as there are
+additional bytes in the encoding, followed by a 0, and the bits of the number. Numbers between 0 and 127 are encoded
+in one byte, and each additional byte adds 7 more bits. Values that use all 8 bytes do not need a 9th bit of 0 and can
+thus fit 9 bytes. Because longer numbers have more 1s in their MSBs, they compare
+higher than shorter ones (and we always use the shortest representation). Because the length is specified through these
+initial bits, no value can be a prefix of another.
+
+
+| Value            | bytes                   | encodes as                 |
+| ---------------- | ----------------------- | -------------------------- |
+| 0                | 00 00 00 00 00 00 00 00 | 00                         |
+| 1                | 00 00 00 00 00 00 00 01 | 01                         |
+| 127 (2^7-1)      | 00 00 00 00 00 00 00 7F | 7F                         |
+| 128 (2^7)        | 00 00 00 00 00 00 00 80 | 80 80                      |
+| 16383 (2^14 - 1) | 00 00 00 00 00 00 3F FF | BF FF                      |
+| 16384 (2^14)     | 00 00 00 00 00 00 40 00 | C0 40 00                   |
+| 2^31 - 1         | 00 00 00 00 7F FF FF FF | F0 7F FF FF FF             |
+| 2^31             | 00 00 00 00 80 00 00 00 | F0 80 00 00 00             |
+| 2^56 - 1         | 00 FF FF FF FF FF FF FF | FE FF FF FF FF FF FF FF    |
+| 2^56             | 01 00 00 00 00 00 00 00 | FF 01 00 00 00 00 00 00 00 |
+| 2^64- 1          | FF FF FF FF FF FF FF FF | FF FF FF FF FF FF FF FF FF |
```

Comment:
Asking out of curiosity, how did you generate these bytes and encoded as lists? I wanted to understand more, but I couldn't seem to reproduce the mapping. I wrote this little test, and it produced the below mapping:

```java
    @Test
    public void testRepresentation()
    {
        printMe(BigInteger.ZERO);
        printMe(BigInteger.ONE);
        printMe(BigInteger.TWO.pow(7).subtract(BigInteger.ONE));
        printMe(BigInteger.TWO.pow(7));
        printMe(BigInteger.TWO.pow(14).subtract(BigInteger.ONE));
        printMe(BigInteger.TWO.pow(14));
        printMe(BigInteger.TWO.pow(31).subtract(BigInteger.ONE));
        printMe(BigInteger.TWO.pow(31));
        printMe(BigInteger.TWO.pow(56).subtract(BigInteger.ONE));
        printMe(BigInteger.TWO.pow(56));
        printMe(BigInteger.TWO.pow(64).subtract(BigInteger.ONE));
        printMe(BigInteger.TWO.pow(64));
    }

    private void printMe(BigInteger val)
    {
        var valBB = IntegerType.instance.getSerializer().serialize(val);
        StringBuilder sb = new StringBuilder();
        for (byte b : valBB.array()) {
            sb.append(String.format("%02X ", b));
        }
        sb.append(" -> ");
        var byteSource = IntegerType.instance.asComparableBytes(ByteBufferAccessor.instance, valBB, Version.OSS50);
        var bytes = ByteSourceInverse.readBytes(byteSource);
        for (byte b : bytes) {
            sb.append(String.format("%02X ", b));
        }
        System.out.println(sb);
    }
```

```
00  -> 80 
01  -> 81 
7F  -> C0 7F 
00 80  -> C0 80 
3F FF  -> E0 3F FF 
40 00  -> E0 40 00 
7F FF FF FF  -> F8 7F FF FF FF 
00 80 00 00 00  -> F8 80 00 00 00 
00 FF FF FF FF FF FF FF  -> FF 00 FF FF FF FF FF FF FF 
01 00 00 00 00 00 00 00  -> FF 01 01 00 00 00 00 00 00 00 
00 FF FF FF FF FF FF FF FF  -> FF 01 FF FF FF FF FF FF FF FF 
01 00 00 00 00 00 00 00 00  -> FF 02 01 00 00 00 00 00 00 00 00 
```

Reply:
What you got with this code is the signed version, specifically in the varint flavour, which is described a few sections below (starting at line 562). The table above is constructed using `ByteSource.VariableLengthUnsignedInteger`, and the one below using `ByteSource.VariableLengthInteger`. The former is only used as part of the varint encoding, the latter both for shorter varints and bigint.